be/src/formats/orc/orc_chunk_writer.cpp

```diff
@@ -0,0 +1,739 @@
+        break;
+    }
+    case TYPE_SMALLINT: {
+        _write_numbers<TYPE_SMALLINT, orc::LongVectorBatch>(orc_column, column);
+        break;
+    }
+    case TYPE_INT: {
+        _write_numbers<TYPE_INT, orc::LongVectorBatch>(orc_column, column);
+        break;
+    }
+    case TYPE_BIGINT: {
+        _write_numbers<TYPE_BIGINT, orc::LongVectorBatch>(orc_column, column);
+        break;
+    }
+    case TYPE_FLOAT: {
+        _write_numbers<TYPE_FLOAT, orc::DoubleVectorBatch>(orc_column, column);
+    }
+    case TYPE_DOUBLE: {
+        _write_numbers<TYPE_DOUBLE, orc::DoubleVectorBatch>(orc_column, column);
+    }
+    case TYPE_CHAR:
+        [[fallthrough]];
+    case TYPE_VARCHAR: {
+        _write_strings(orc_column, column);
+        break;
+    }
+    case TYPE_DECIMAL: {
+        _write_decimals(orc_column, column, type_desc.precision, type_desc.scale);
+        break;
+    }
+    case TYPE_DECIMAL32: {
+        _write_decimal32or64or128<TYPE_DECIMAL32, orc::Decimal64VectorBatch, int64_t>(
+                orc_column, column, type_desc.precision, type_desc.scale);
+        break;
+    }
+    case TYPE_DECIMAL64: {
+        _write_decimal32or64or128<TYPE_DECIMAL64, orc::Decimal64VectorBatch, int64_t>(
+                orc_column, column, type_desc.precision, type_desc.scale);
+        break;
+    }
+    case TYPE_DECIMAL128: {
+        _write_decimal32or64or128<TYPE_DECIMAL128, orc::Decimal128VectorBatch, int128_t>(
+                orc_column, column, type_desc.precision, type_desc.scale);
+        break;
+    }
+    case TYPE_DATE: {
+        _write_datetimes(orc_column, column);
+        break;
+    }
+    case TYPE_DATETIME: {
+        _write_timestamps(orc_column, column);
+        break;
+    }
+    case TYPE_ARRAY: {
+        RETURN_IF_ERROR(_write_array_column(orc_column, column, type_desc));
+        break;
+    }
+    case TYPE_STRUCT: {
+        RETURN_IF_ERROR(_write_struct_column(orc_column, column, type_desc));
+        break;
+    }
+    case TYPE_MAP: {
+        RETURN_IF_ERROR(_write_map_column(orc_column, column, type_desc));
+        break;
+    }
+    default:
+        return Status::NotSupported(strings::Substitute("Type $0 not supported", type_desc.type));
+    }
+    return Status::OK();
+}
+
+template <LogicalType Type, typename VectorBatchType>
+void OrcChunkWriter::_write_numbers(orc::ColumnVectorBatch& orc_column, ColumnPtr& column) {
+    auto& number_orc_column = dynamic_cast<VectorBatchType&>(orc_column);
+    auto column_size = column->size();
+
+    number_orc_column.resize(column_size);
+    number_orc_column.notNull.resize(column_size);
+
+    if (column->is_nullable()) {
+        auto c = ColumnHelper::as_raw_column<NullableColumn>(column);
+        auto* nulls = c->null_column()->get_data().data();
+        auto* values = ColumnHelper::cast_to_raw<Type>(c->data_column())->get_data().data();
+
+        for (size_t i = 0; i < column_size; ++i) {
+            if (nulls[i]) {
```

Comment:
Is it possible to remove this in-loop branch (to enable auto-vectorization)? 

I guess it's ok to populate `number_orc_column[i]` even if `nulls[i] != 0`.

Reply:
Do you mean do redundant copy for auto-vectorization? I think it is an good idea and have changed.